### ISSUE TITLE
Deprecate "Default" Environment classes and NodesEnvironment

### ISSUE THREAD

**Issue by bdice** [glotzerlab/signac-flow#492]
Currently we have a significant class hierarchy that serves no apparent purpose in signac-flow's environments. Here's a simplified look:

```python
class SlurmEnvironment(ComputeEnvironment):
    scheduler_type = SlurmScheduler
    template = "slurm.sh"

class NodesEnvironment(ComputeEnvironment):
    pass  # only has a docstring

class DefaultSlurmEnvironment(SlurmEnvironment, NodesEnvironment):
    # docstring

    @classmethod
    def add_args(cls, parser):
        # method contents...
        pass
```

I think it would be good to reduce this hierarchy and only define:

```python
class DefaultSlurmEnvironment(ComputeEnvironment):
    # docstring
    scheduler_type = SlurmScheduler
    template = "slurm.sh"

    @classmethod
    def add_args(cls, parser):
        # method contents...
        pass
```

This pattern exists for all three supported schedulers. It would take a deprecation cycle to change this, but I don't think that I can explain the rationale for the current hierarchy and it provides no clear benefit. I would probably keep the name `DefaultSlurmEnvironment` instead of `SlurmEnvironment` since that's what external users are more likely to have used as a base class for their own environments and keeping the word "Default" in the class name is a small price to pay for that compatibility.

Recently @atravitz has asked me why certain things are organized as they are in signac-flow, and I think that simplifying this is a step in the right direction so that future users can quickly understand how the software works. Tagging @csadorf since he authored most of this code and might have a perspective I'm missing.

**Comment by csadorf** [comment]
>  Tagging @csadorf since he authored most of this code and might have a perspective I'm missing.

The class hierarchy was designed to avoid any premature specialization since it is always possible to add new class members vie inheritance, but impossible to remove them. This is the motivation for the introduction of the `NodesEnvironment`, which makes assumption about the fact that the particular environment *has nodes* and I assume at some point defined (directive) attributes relating to that. This is probably a premature generalization, because we could in principle introduce another base class should we encounter an environment where this is not true. So I agree that this class can probably be dropped without any replacement.

The `SlurmEnvironment` (as opposed to `SlurmDefaultEnvironment`) is a base class from which all other SLURM environments should be derived. It was never intended to be instantiated directly, because at the time *all* environments were specialized. It should probably have been named `SlurmBaseEnvironment` or so right from the start. We then later introduced the `DefaultSlurmEnvironment` as a fallback environment that might work when there is no more specialized environment. If it can be shown that the base class can be dropped without introducing abstraction issues, I'm all for it.

**Comment by bdice** [comment]
Resolved by #494.
